### ISSUE TITLE
Multiply the two, and set it in perturbation?

### ISSUE THREAD

**Issue by HenryPengZou** [eth-sri/probabilistic-forecasts-attacks#5]
https://github.com/eth-sri/probabilistic-forecasts-attacks/blob/df78ffd7375089bb7b38fbf55bb840b7699365eb/Finance/attack/attacks.py#L104-L109

Hi, what does it mean here to 'Multiply the two, and set it in perturbation', could you provide a mathematical description? I think it has something to do with two gradients.

**Comment by rdang-nhu** [comment]
Hi. The `attack_step_score` method is the implementation of **Score function estimator 3.1** in [the paper](https://arxiv.org/pdf/2003.03778.pdf). See #4 for the meaning of the different quantities. This multiplication is essentially a manual application of the chain rule, where the gradient loss -> mean (computed with `loss.backward()`) and the gradient mean -> perturbation (computed with `aux_estimate.backward()`). This application of the chain rule needs to be done manually because score function estimators require to freeze different gradients at different steps (with `requires_grad = False`).

<img width="991" alt="Screenshot 2021-07-23 at 10 43 09" src="https://user-images.githubusercontent.com/18661420/126758053-7ed84134-09fb-49ca-b24e-d0b01b12f2a4.png">

**Comment by HenryPengZou** [comment]
Thanks for your excellent answer!
